File: 07-ruby/game.rb
# Mantiene la lógica y el estado de una partida.
class Game
  NUMERO_MINIMO = 1
  NUMERO_MAXIMO = 100

  attr_reader :intentos

  def initialize
    reiniciar
  end

  def comprobar_numero(numero)
    return '¡Correcto!' if terminado?

    unless numero.is_a?(Integer) && numero.between?(NUMERO_MINIMO, NUMERO_MAXIMO)
      raise ArgumentError, 'El número debe estar entre 1 y 100.'
    end

    @intentos += 1

    if numero < @numero_secreto
      'Muy bajo'
    elsif numero > @numero_secreto
      'Muy alto'
    else
      @terminado = true
      '¡Correcto!'
    end
  end

  def reiniciar
    @numero_secreto = rand(NUMERO_MINIMO..NUMERO_MAXIMO)
    @intentos = 0
    @terminado = false
  end

  def terminado?
    @terminado
  end
end

File: 08-go/game.go
package main

import "math/rand/v2"

const (
	NumeroMinimo = 1
	NumeroMaximo = 100
)

// Juego mantiene la lógica y el estado de una partida.
type Juego struct {
	numeroSecreto int
	Intentos      int
	finalizado    bool
}

func NuevoJuego() *Juego {
	juego := &Juego{}
	juego.Reiniciar()
	return juego
}

func (j *Juego) ComprobarNumero(numero int) string {
	if j.finalizado || numero < NumeroMinimo || numero > NumeroMaximo {
		return ""
	}

	j.Intentos++

	switch {
	case numero < j.numeroSecreto:
		return "Muy bajo"
	case numero > j.numeroSecreto:
		return "Muy alto"
	default:
		j.finalizado = true
		return "¡Correcto!"
	}
}

func (j *Juego) Reiniciar() {
	j.numeroSecreto = rand.IntN(NumeroMaximo-NumeroMinimo+1) + NumeroMinimo
	j.Intentos = 0
	j.finalizado = false
}

func (j *Juego) Finalizado() bool {
	return j.finalizado
}
